velox/exec/ExchangeAggregation.h

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <folly/CppAttributes.h>
+
+#include "velox/exec/GroupingSet.h"
+#include "velox/exec/Operator.h"
+#include "velox/exec/RowContainer.h"
+
+namespace facebook::velox::exec {
+
+struct Partitioned {
+  RowVectorPtr input;
+  BufferPtr partitions;
+  std::vector<vector_size_t> offsets;
+  uint64_t size;
+  std::atomic<int32_t> accessCount{0};
+
+  Partitioned(
+      const RowVectorPtr& input,
+      const BufferPtr& partitions,
+      const std::vector<vector_size_t>& offsets,
+      uint64_t size)
+      : input(input), partitions(partitions), offsets(offsets), size(size) {}
+
+  Partitioned() = delete;
+  Partitioned(const Partitioned&) = delete;
+  Partitioned(Partitioned&&) = delete;
+
+  ~Partitioned() {
+    input = nullptr;
+    partitions = nullptr;
+  }
+};
+
+class PartitionBucket {
+ public:
+  PartitionBucket(size_t bucketSize, int32_t driverCount);
+
+  bool tryAdd(Partitioned* partitioned);
+
+  Partitioned* FOLLY_NULLABLE getSlotAt(int32_t index);
+
+  void noMoreInput() {
+    noMoreinputDriverCount_.fetch_add(1);
+  }
+
+  bool isFinished() {
+    if (noMoreinputDriverCount_.load() == driverCount_) {
+      if (partitionCount_.load() == 0) {
+        return true;
+      } else {
+        needToConsume_.store(true);
+      }
+    }
+    return false;
+  }
+
+  size_t bucketSize() const {
+    return bucketSize_;
+  }
+
+  int32_t partitionCount() const {
+    return partitionCount_.load();
+  }
+
+  uint64_t memoryUsage() const {
+    return memoryUsage_.load();
+  }
+
+  inline void requestToConsume() {
+    needToConsume_.store(true);
+  }
+
+  inline void resetNeedToConsume() {
+    needToConsume_.store(false);
+  }
+
+  inline bool needToConsume() const {
+    return needToConsume_.load();
+  }
+
+  ~PartitionBucket() = default;
+
+  void testingValidate();
+
+ private:
+  size_t bucketSize_;
+  std::vector<std::atomic<Partitioned*>> slots_;
```

Comment:
Double check each element is aligned at cache line size (64 bytes) to avoid false sharing.  You may need to wrap the `Partitioned*` in a struct annotated with `alignas(64)` to fix this.